mindsdb_native/libs/helpers/locking.py

```diff
@@ -1,34 +1,77 @@
 import os
-
+import sys
+import inspect
+import traceback
 import portalocker
 
 from mindsdb_native.config import CONFIG
 
 
-def learn_lock(name):
-    learn_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_learn.lock')
-    f = open(learn_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_EX+portalocker.LOCK_NB)
-    return f
-
-def predict_lock(name):
-    learn_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_learn.lock')
-    f = open(learn_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_SH+portalocker.LOCK_NB)
-    return f
-
-def delete_lock(name):
-    learn_lock(name)
-    delete_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_delete.lock')
-    f = open(delete_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_EX+portalocker.LOCK_NB)
-    return f
-
-def get_data_lock(name):
-    delete_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_delete.lock')
-    f = open(delete_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_SH+portalocker.LOCK_NB)
-    return f
-
-def unlock(file):
-    portalocker.unlock(file)
+def mdb_lock(flags, lock_name, argname=None):
+    """
+    :param lock_name: str, name of the lock
+    :argname: str, name of the positional/keyword argument of
+    the input function that will be concatenated with :lock_name:
+    """
+    def wrapper1(func):
+        def wrapper2(*args, **kwargs):
+            if argname is None:
+                final_lock_name = '{}.lock'.format(lock_name)
+            else:
+                if argname.startswith('self.'):
+                    argval = getattr(args[0], argname.split('.')[1])
+                elif argname in kwargs:
+                    argval = kwargs[argname]
+                else:
+                    index = inspect.getfullargspec(func).args.index(argname)
+                    if index < len(args):
+                        argval = args[index]
+                    else:
+                        raise Exception('argname wasn\'t found in *args/**kwargs')
```

Comment:
I fail to see why we need this bit of logic. Why not make `argname` mandatory and just name the lock `{lock_name}_{argname}.lock`.

After all `argname` basically ought to be the name of the Predictor we are locking, that should be a name we always have access to (maybe even rename it to this)

Reply:
ahhh... nvm, I just realized, this is because the decorator is external to the scope :(

Well, for once, I think you should removed the None check and not allow it to be none, but second, this makes things a bit complicated.

Are you sure the logic for getting `argname` here always works ? If not, would it be hard to replace with decorator with a `with` in order to simplify the logic ?

Reply:
1. "Well, for once, I think you should removed the None check and not allow it to be none, but second, this makes things a bit complicated."

Allowing it to be `None` might be useful if we for example want to lock Predictor.learn no matter what Predictor.name is equal to.

2. If we use `with` statement, then we have to do it everywhere. The goal of this decorator is to be able to just put it over any method or function, and whole locking logic is abstracted away.

3. Getting function argument from withing the decorator by the name of the argument is kind of tricky, I'm sure it's possible to do (with `inspect` module) but I'm not 100% certain that this implementation will not ever fail. We need some tests for it.

Reply:
> Allowing it to be None might be useful if we for example want to lock Predictor.learn no matter what Predictor.name is equal to.

We will never have to do this, but we might by mistake, hence why I'm saying None should be removed

> If we use with statement, then we have to do it everywhere. The goal of this decorator is to be able to just put it over any method or function, and whole locking logic is abstracted away.

Fair enough, seem like a preference between indentation vs having to use function declarations inside a function declaration ,whichever works, I'm fine sticking with the decorator.

So, if you can get the tests rolling, prefer. You only need to test with the function names for which it is used.

Reply:
Ah, nevermind, I see you change to a `with` now :P